lib/spack/spack/build_systems/cached_cmake.py

```diff
@@ -257,6 +262,15 @@ def initconfig_hardware_entries(self):
             entries.append(
                 cmake_cache_path("HIP_CXX_COMPILER", "{0}".format(self.spec["hip"].hipcc))
             )
+            llvm_bin = spec["llvm-amdgpu"].prefix.bin
+            llvm_prefix = spec["llvm-amdgpu"].prefix
+            # Some ROCm systems seem to point to /<path>/rocm-<ver>/ and
+            # others point to /<path>/rocm-<ver>/llvm
+            if os.path.basename(os.path.normpath(llvm_prefix)) != "llvm":
+                llvm_bin = os.path.join(llvm_prefix, "llvm/bin/")
+            entries.append(
+                cmake_cache_filepath("CMAKE_HIP_COMPILER", os.path.join(llvm_bin, "clang++"))
+            )
```

Comment:
@bvanessen do you remember the exact explanation for this ?
I have been in the middle of a revamp in #39613 and now in #41375 instead, and I am trying to find the right way of supporting all the cases.
My understanding is that we need this for recent rocm builds, when the spack spec is asking for +rocm without a proper amdgpu compiler.
However, this is co-existing with the "old fashion way", which is to set HIP_CXX_COMPILER to hipcc (see line 263) and I wonder in it should be the case.